File: src/python/problems/combination_sum_3_test.py
# https://leetcode.com/problems/combination-sum-iii

from core.problem_base import *
import collections

class CombinationSum3(ProblemBase):
    def Solution(self, k: int, n: int) -> list[list[int]]:
        result = []
        edge = min(9, n - k + 1)

        def search(current, temp):
            if len(temp) == k:
                if sum(temp) == n:
                    result.append(list(temp))

                return

            for i in range(current, edge + 1):
                temp.append(i)
                search(i + 1, temp)
                temp.pop()

        search(1, [])

        return result


if __name__ == '__main__':
    TestGen(CombinationSum3) \
        .Add(lambda tc: tc.Param("k", 3).Param("n", 7).Result([[1,2,4]])) \
        .Add(lambda tc: tc.Param("k", 3).Param("n", 9).Result([[1,2,6],[1,3,5],[2,3,4]])) \
        .Add(lambda tc: tc.Param("k", 4).Param("n", 1).Result([])) \
        .Run()

File: src/go/problems/combination_sum_3_test.go
// https://leetcode.com/problems/combination-sum-iii

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type CombinationSum3 struct{}

func TestCombinationSum3(t *testing.T) {
	gen := core.TestSuite[CombinationSum3]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(3).Param(7).Result([][]int{{1,2,4}})
	}).Add(func(tc *core.TestCase) {
		tc.Param(3).Param(9).Result([][]int{{1,2,6},{1,3,5},{2,3,4}})
	}).Add(func(tc *core.TestCase) {
		tc.Param(4).Param(1).Result([][]int{})
	}).Run(t)
}

func (CombinationSum3) Solution(k int, n int) [][]int {
    result := [][]int{}

    var search func(int, int, *[]int) 
    search = func(current int, sum int, temp *[]int) {
        if len(*temp) == k {
            if sum == n {
                result = append(result, append([]int{}, (*temp)...))
            }

            return
        }

        for i := current; i <= 9; i++ {
            *temp = append(*temp, i)
            search(i + 1, sum + i, temp)
            *temp = (*temp)[:len(*temp)-1]
        }
    }

    search(1, 0, &[]int{})

    return result
}
